src/operations/client_bulk_write/command_builder.ts

```diff
@@ -0,0 +1,302 @@
+  model: ClientDeleteOneModel | ClientDeleteManyModel,
+  index: number,
+  multi: boolean
+): Document {
+  const document: Document = {
+    delete: index,
+    multi: multi,
+    filter: model.filter
+  };
+  if (model.hint) {
+    document.hint = model.hint;
+  }
+  if (model.collation) {
+    document.collation = model.collation;
+  }
+  return document;
+}
+
+/** @internal */
+export class UpdateOneOperationBuilder implements OperationBuilder {
+  model: ClientUpdateOneModel;
+
+  /**
+   * Instantiate the builder.
+   * @param model - The client update one model.
+   */
+  constructor(model: ClientUpdateOneModel) {
+    this.model = model;
+  }
+
+  /**
+   * Build the operation.
+   * @param index - The namespace index.
+   * @returns the operation.
+   */
+  buildOperation(index: number): Document {
+    return createUpdateOperation(this.model, index, false);
+  }
+}
+
+/** @internal */
+export class UpdateManyOperationBuilder implements OperationBuilder {
+  model: ClientUpdateManyModel;
+
+  /**
+   * Instantiate the builder.
+   * @param model - The client update many model.
+   */
+  constructor(model: ClientUpdateManyModel) {
+    this.model = model;
+  }
+
+  /**
+   * Build the operation.
+   * @param index - The namespace index.
+   * @returns the operation.
+   */
+  buildOperation(index: number): Document {
+    return createUpdateOperation(this.model, index, true);
+  }
+}
+
+/**
+ * Creates a delete operation based on the parameters.
+ */
+function createUpdateOperation(
+  model: ClientUpdateOneModel | ClientUpdateManyModel,
+  index: number,
+  multi: boolean
+): Document {
+  const document: Document = {
+    update: index,
+    multi: multi,
+    filter: model.filter,
+    updateMods: model.update
+  };
+  if (model.hint) {
+    document.hint = model.hint;
+  }
+  if (model.upsert) {
+    document.upsert = model.upsert;
+  }
+  if (model.arrayFilters) {
+    document.arrayFilters = model.arrayFilters;
+  }
+  return document;
+}
+
+/** @internal */
+export class ReplaceOneOperationBuilder implements OperationBuilder {
+  model: ClientReplaceOneModel;
+
+  /**
+   * Instantiate the builder.
+   * @param model - The client replace one model.
+   */
+  constructor(model: ClientReplaceOneModel) {
+    this.model = model;
+  }
+
+  /**
+   * Build the operation.
+   * @param index - The namespace index.
+   * @returns the operation.
+   */
+  buildOperation(index: number): Document {
+    const document: Document = {
+      update: index,
+      multi: false,
+      filter: this.model.filter,
+      updateMods: this.model.replacement
+    };
+    if (this.model.hint) {
+      document.hint = this.model.hint;
+    }
+    if (this.model.upsert) {
+      document.upsert = this.model.upsert;
+    }
+    return document;
+  }
+}
+
+const BUILDERS: Map<string, (model: AnyClientBulkWriteModel) => OperationBuilder> = new Map();
+BUILDERS.set('insertOne', model => new InsertOneOperationBuilder(model as ClientInsertOneModel));
```

Comment:
If you want to avoid type-casting the model - the typical pattern in TS is to use a switch statement on the discriminator in your union type, which TS uses to narrow the type for you:

```typescript
export function builder(model: AnyClientBulkWriteModel): OperationBuilder {
  switch (model.name) {
    case 'insertOne':
      return new InsertOneOperationBuilder(model);
    // ... TS can auto-generate all possible cases for the switch
    default:
      throw new MongoInvalidArgumentError(`Could not load builder for model ${model.name}`);
  }
}
```

Reply:
I've updated to use a switch on the new functions now.